scalyr_agent/sharded_copying_manager/worker.py

```diff
@@ -1028,19 +1025,271 @@ def __write_full_checkpoint_state(self, current_time):
+
+
+class CopyingManagerWorkerContainer(CopyingManagerWorker):
+    """
+    A container/wrapper class for the real CopyingManagerThreadedWorker class.
+    This container initializes the real worker according to the configuration.
+    If the type of the worker container is "thread", then a simple threaded worker is created.
+    If the type is "process" - still the same  CopyingManagerThreadedWorker is created, but inside a separate process.
+    The communication with the worker from the process if made by multiprocessing.SyncManager.
+    This shared memory manager starts its own process and will create worker instance within it.
+    It also provides a proxy object of the worker, so we can interact with it from the main process.
+    We also create proxy objects for LogFileProcessor instances, because log processors should be located incide worker,
+     but we still have to be able to interact with them
+    """
+    # methods of the CopyingManagerThreadedWorker which should be exposed through proxies.
+    EXPOSED = [
+        six.ensure_str("start_worker"),
+        six.ensure_str("stop_worker"),
+        six.ensure_str("wait_for_copying_to_begin"),
+        six.ensure_str("get_id"),
+        six.ensure_str("augment_user_agent_for_client_session"),
+        six.ensure_str("generate_status"),
+        six.ensure_str("generate_scalyr_client_status"),
+        six.ensure_str("schedule_new_log_processor"),
+        six.ensure_str("generate_scalyr_client_status"),
+        six.ensure_str("get_log_processors")
+    ]
+
+    # Class of the worker that should be proxified.
+    # This class attribute is needed only to be able to replace this class by the testable one,
+    # just for the testing purposes.
+    COPYING_MANAGER_CLASS = CopyingManagerThreadedWorker
+
+    def __init__(self, configuration, api_key_config_entry, worker_id):
+
+        copying_manager_class = type(self)
+        self._worker_type = api_key_config_entry.get("type", none_if_missing=True)
+        if self._worker_type == "process":
+            # the type of the real worker is a process. Initialize shared memory manager and register proxy types.
+
+            # the proxy types are registered for the whole class, so we create a new subclass to avoid mixing.
+            class CopyingManagerMemoryManager(SyncManager):
+                pass
+
+
+            # register Proxy type for the worker
+            CopyingManagerMemoryManager.register(
+                six.ensure_str("CopyingManagerWorker"),
+                copying_manager_class.COPYING_MANAGER_CLASS,
+                exposed=copying_manager_class.EXPOSED,
+                method_to_typeid={six.ensure_str("get_log_processors"): six.ensure_str("list")}
+            )
+
+            # register proxy type for the LogFilePRocessor
+            CopyingManagerMemoryManager.register(
+                six.ensure_str("LogFileProcessorProxy"),
+                LogFileProcessor,
+                exposed=[
```

Comment:
Are these supposed to be all of the method that can be invoked across the process boundary?  This is what the copy manager would invoke, is that right?

If so, I'm a little surprised.  I would think that the only ones that need to be invoked across process boundaries are `close_at_eof` and `close`.  Maybe `add_missing_attributes`.  I have to dig into that one a little more.  But, the worker (which is in the same process as the processor) should be handling the checkpointing and the status generation, I would think.

Reply:
Yes, I removed the unneeded methods. Now there are only:
*  close_at_eof
* close
* add_missing_attributes
* get_log_path # this one does not change, so I made it cacheable on the proxy side.
* generate_status # used by LogMatcher's

Reply:
Update: also `is_closed` method is invoked externally by the LogMatcher.